### ISSUE TITLE
Need to adjust handling timeouts during/after machine deletion

### ISSUE THREAD

**Issue by ncdc** [vmware/cluster-api-upgrade-tool#121]
Our current logic is:

1. List all non-deleted control plane machines
1. For each machine
    1. Do all the work required to create a replacement Machine, etc.
    1. Remove original from etcd
    1. Delete original Machine
    1. Wait for deletion to finish
    1. Remove original node's entry from kubeadm configmap

If we get a timeout waiting for the deleted Machine to fully go away, or trying to update the kubeadm configmap, we won't ever reprocess the deleted Machine the next time the tool runs.

To correct this, we should:

1. List all machines
1. For each machine
    1. If it is active, perform the replacement logic (not including deletion)
    1. Regardless of status, perform deletion logic + configmap updating
        1. If deletion fails with a NotFound error, that's ok, proceed to configmap update

There is still a chance we could miss something in between upgrade runs. For example:

1. Try to delete machine - get a timeout
1. Wait some amount of time - machine gets actually deleted
1. Rerun tool - because machine is gone, we don't see it again, and and miss removing it from the kubeadm configmap

To mitigate, at the beginning and/or end of each invocation, we could update the kubeadm configmap so it matches the nodes that currently exist when we run `kubectl get nodes` against the workload cluster.

cc @chuckha @rudoi @williamsandrew for 👀 and feedback

**Comment by ncdc** [comment]
and cc @detiber @vincepri

**Comment by ncdc** [comment]
cc @imikushin

**Comment by detiber** [comment]
I think we should likely be cautious about deleting an etcd member if we do not know if the deletion is successful. If the Machine is still responding, it could cause issues for any apiserver that is still configured with that host as an etcd endpoint. Outside of that, I agree with the rest of the workflow.

**Comment by ncdc** [comment]
You're correct, but I don't think we can do any better, given the state of v1alpha2. I think the ideal sequence would be:

1. Create replacement machine, wait for it to be ready
1. Remove old machine from load balancer (drain traffic)
1. Remove old machine from etcd membership
1. Delete old machine
1. Remove old machine from kubeadm configmap

We aren't able to control the timing of step 2, other than by deleting the machine (which has the side effect, at least for a provider like CAPA, of removing the machine from the load balancer).

Do you think there is anything we can do today to help here? If not, I think we're going to have to keep it as

1. Create replacement machine
1. Delete from etcd
1. Delete original machine
1. Update kubeadm configmap

**Comment by ncdc** [comment]
Or we could swap, and delete the machine first before we delete from etcd, but that might be worse (or equally not great)?

**Comment by vincepri** [comment]
Upon deletion, the Machine should be cordoned and drained. Does drain also affect control plane resources? If so, we can probably assume that when we remove the Machine infrastructure, the health check will be failing and the machine won't be responding to requests anymore.

**Comment by detiber** [comment]
Might need to think a bit more about the etcd membership ordering...  

For the kubeadm config map, I'm wondering if it might be good to go ahead and remove the "old" machine prior to bringing up the new machine. That would prevent the "new" Machine from having any configuration references to the "old" Machine. This would result in the new instance not having the old etcd instance configured as a client or a peer, but I think that *should* be ok...  Would need to test it out, though.

**Comment by detiber** [comment]
> Upon deletion, the Machine should be cordoned and drained. Does drain also affect control plane resources? If so, we can probably assume that when we remove the Machine infrastructure, the health check will be failing and the machine won't be responding to requests anymore.

Draining and cordoning will not affect any static pods

**Comment by chuckha** [comment]
> For the kubeadm config map, I'm wondering if it might be good to go ahead and remove the "old" machine prior to bringing up the new machine. That would prevent the "new" Machine from having any configuration references to the "old" Machine. This would result in the new instance not having the old etcd instance configured as a client or a peer, but I think that _should_ be ok... Would need to test it out, though.

This won't work for single node control plane clusters, but maybe we could consider two different workflows

**Comment by ncdc** [comment]
I like the idea of manipulating the kubeadm configmap earlier as you suggested. I also think we should have a "pre" step at the beginning of an upgrade that compares the actual nodes w/the entries of the configmap and makes sure the configmap matches what's there. And we should do the same thing at the end of the upgrade process, too.

**Comment by ncdc** [comment]
Oh right, single nodes...

**Comment by detiber** [comment]
We could potentially poll the etcd endpoint for the machine we are deleting and wait for it to no longer respond before removing the etcd membership... That could potentially cause issues when going from 2->1 though as the cluster would lose quorum...  :(

**Comment by chuckha** [comment]
+1 to the idea of building state at the start, making sure it all makes sense and then upgrading. Andy and I talked about this idea a bit a while back, but it seemed like too big of a departure from what we were doing. Maybe this is enough evidence that we should move to a more reconcile-like upgrade tool?

**Comment by ncdc** [comment]
Oh, I'm not suggesting that we need to go all-in right now on state - just the nodes vs the kubeadm configmap data.

**Comment by ncdc** [comment]
Thinking about this some more, I wonder if instead of doing what I proposed in the description, we instead just add doing a full reconcile of the kubeadm configmap at both the beginning and the end of the upgrade process. If you try to do an upgrade and it makes it all the way through without errors, it's all good. If there's a timeout somewhere and you need to resume the upgrade, about the only place we potentially can miss out on something is if we fail to remove a now-delete machine from the kubeadm configmap. And adding a full reconcile at the beginning and the end of the upgrade process should close that gap. WDYT?

**Comment by detiber** [comment]
I like the idea of reconciling the configmap at both start/end
